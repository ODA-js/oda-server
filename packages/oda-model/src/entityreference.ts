import * as inflected from 'inflected';
import decapitalize from './lib/decapitalize';
import { merge } from 'lodash';

import { DEFAULT_ID_FIELDNAME, REF_PATTERN } from './definitions';
import { MetaModelType, Nullable, assignValue } from './types';
import {
  ElementMetaInfo,
  ElementInternal,
  Element,
  ElementInput,
  ElementOutput,
  IUpdatable,
} from './element';

function StrToEntityRef(input: string): EntityRefInput {
  let res = input.match(REF_PATTERN);
  return {
    backField: decapitalize(res && res[1] ? res[1] : ''),
    entity: inflected.classify(res && res[2] ? res[2] : ''),
    field: decapitalize(res && res[3] ? res[3].trim() : DEFAULT_ID_FIELDNAME),
  };
}

export interface IEntityRef
  extends IUpdatable<EntityRefMetaInfo, EntityRefInput, EntityRefOutput> {
  /**
   * referencing entity
   */
  readonly entity: string;
  /**
   * referencing key field
   */
  readonly field: string;
  /**
   * the backed field
   * field that in owner of the reference is storing the key
   */
  readonly backField?: string;
}

export interface EntityRefMetaInfo extends ElementMetaInfo {}

export interface EntityRefInternal extends ElementInternal<EntityRefMetaInfo> {
  entity: string;
  field: string;
  backField?: string;
}

export interface EntityRefInput extends ElementInput<EntityRefMetaInfo> {
  entity: string;
  field?: string;
  backField?: string;
}

export interface EntityRefOutput extends ElementOutput<EntityRefMetaInfo> {
  entity: string;
  backField?: string;
  field: string;
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

/** Entity reference implementation */
export class EntityReference
  extends Element<
    EntityRefMetaInfo,
    EntityRefInput,
    EntityRefInternal,
    EntityRefOutput
  >
  implements IEntityRef {
  public get modelType(): MetaModelType {
    return 'ref';
  }
  /** the Entity that is referenced */
  public get entity(): string {
    return this.$obj.entity;
  }
  /** the Identity field */
  public get field(): string {
    return this.$obj.field;
  }

  /** the Identity field */
  public get backField(): string {
    return this.$obj.backField || '';
  }

  constructor(init: EntityRefInput | string) {
    super(
      merge(
        {},
        defaultInput,
        typeof init === 'object' ? init : StrToEntityRef(init),
      ),
    );
  }

  public toString(): string {
    return `${this.backField ? this.backField + '@' : ''}${this.$obj.entity}#${
      this.$obj.field
    }`;
  }

  public updateWith(input: Nullable<EntityRefInput>) {
    super.updateWith(typeof input === 'object' ? input : StrToEntityRef(input));

    assignValue<EntityRefInternal, EntityRefInput, EntityRefInput['entity']>({
      src: this.$obj,
      input,
      field: 'entity',
      effect: (src, value) => (src.entity = value),
      required: true,
    });

    assignValue<EntityRefInternal, EntityRefInput, EntityRefInput['field']>({
      src: this.$obj,
      input,
      field: 'field',
    });

    assignValue<EntityRefInternal, EntityRefInput, EntityRefInput['backField']>(
      { src: this.$obj, input, field: 'backField' },
    );
  }

  public toObject(): EntityRefOutput {
    return merge({}, super.toObject(), {
      backField: this.$obj.backField,
      entity: this.$obj.entity,
      field: this.$obj.field,
    } as Partial<EntityRefOutput>) as EntityRefOutput;
  }
}
