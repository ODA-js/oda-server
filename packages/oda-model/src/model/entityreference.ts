import * as camelcase from 'camelcase';
import * as inflected from 'inflected';
import decapitalize from './../lib/decapitalize';

import clean from '../lib/json/clean';
import { DEFAULT_ID_FIELDNAME, REF_PATTERN } from './definitions';
import { EntityReferenceInput, IEntityRef, MetaModelType } from './interfaces';

/** Entityt reference implementation */
export class EntityReference implements IEntityRef {
  public modelType: MetaModelType = 'ref';
  protected $obj: {
    backField_: string;
    entity_: string;
    field_: string;
    backField: string;
    entity: string;
    field: string;
  };
  /** the Entity that is referenced */
  public get entity(): string {
    return this.$obj.entity || this.$obj.entity_;
  }
  public set entity(value: string) {
    this.$obj.entity_ = value;
  }
  /** the Identity field */
  public get field(): string {
    return this.$obj.field || this.$obj.field_ || DEFAULT_ID_FIELDNAME;
  }

  public set field(value: string) {
    this.$obj.field_ = value;
  }

  /** the Identity field */
  public get backField(): string {
    return this.$obj.backField || this.$obj.backField_;
  }

  public set backField(value: string) {
    this.$obj.backField_ = value;
  }

  constructor(
    entity: string | EntityReferenceInput,
    field?: string,
    backField?: string,
  ) {
    if (typeof entity === 'string' && !field) {
      let res = entity.match(REF_PATTERN);
      if (res && res.length > 0) {
        this.$obj = {
          backField: res[1],
          backField_: res[1],
          entity: inflected.classify(res[2]),
          entity_: res[2],
          field: camelcase(res[3].trim()),
          field_: camelcase(res[3].trim()) || DEFAULT_ID_FIELDNAME,
        };
      }
    } else if (typeof entity === 'string') {
      this.$obj = {
        backField: backField,
        backField_: backField,
        entity: inflected.classify(entity),
        entity_: entity,
        field: field,
        field_: field || DEFAULT_ID_FIELDNAME,
      };
    } else if (entity instanceof Object) {
      this.$obj = {
        backField: entity.backField,
        backField_: entity.backField,
        entity: inflected.classify(entity.entity),
        entity_: entity.entity,
        field: entity.field,
        field_: entity.field || DEFAULT_ID_FIELDNAME,
      };
    }
    this.$obj.field = this.$obj.field && decapitalize(this.$obj.field);
    this.$obj.field_ = this.$obj.field_ && decapitalize(this.$obj.field_);

    this.$obj.backField =
      this.$obj.backField && decapitalize(this.$obj.backField);
    this.$obj.backField_ =
      this.$obj.backField_ && decapitalize(this.$obj.backField_);
  }

  public clone() {
    return new (<typeof EntityReference>this.constructor)(this);
  }

  public toObject() {
    return clean({
      backField: this.$obj.backField,
      entity: this.$obj.entity,
      field: this.$obj.field,
    });
  }

  public updateWith(obj: EntityReferenceInput) {
    this.backField = obj.backField || this.backField;
    this.field = obj.field || this.field;
    this.entity = obj.entity || this.entity;
  }

  public toString(): string {
    return `${this.backField ? this.backField + '@' : ''}${this.entity}#${this
      .field || DEFAULT_ID_FIELDNAME}`;
  }
}
