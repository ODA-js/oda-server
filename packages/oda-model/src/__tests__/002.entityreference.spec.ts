import 'jest';
import { EntityReference } from '../entityreference';

describe('EntityReference', () => {
  it('empty string', () => {
    const ref = new EntityReference('');
    expect(ref).toMatchSnapshot();
  });
  it('creates from full string', () => {
    const ref = new EntityReference('me@A#id');
    expect(ref.backField).toBe('me');
    expect(ref.entity).toBe('A');
    expect(ref.field).toBe('id');
  });
  it('creates from short string', () => {
    const ref = new EntityReference('A#');
    expect(ref.backField).toBe('');
    expect(ref.entity).toBe('A');
    expect(ref.field).toBe('id');
  });
  it('creates from short string', () => {
    const ref = new EntityReference('A');
    expect(ref.backField).toBe('');
    expect(ref.entity).toBe('A');
    expect(ref.field).toBe('id');
  });
  it('creates from short string', () => {
    const ref = new EntityReference('me@A#');
    expect(ref.backField).toBe('me');
    expect(ref.entity).toBe('A');
    expect(ref.field).toBe('id');
  });
  it('creates from short object', () => {
    const ref = new EntityReference({ entity: 'A' });
    expect(ref.backField).toBe('');
    expect(ref.entity).toBe('A');
    expect(ref.field).toBe('id');
  });
  it('toString', () => {
    const ref = new EntityReference({ entity: 'A' });
    expect(ref.toString()).toBe('A#id');
  });
});
