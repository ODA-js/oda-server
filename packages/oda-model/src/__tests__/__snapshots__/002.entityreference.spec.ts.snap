// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityReference empty string 1`] = `
EntityReference {
  Symbol(internal): Object {
    "entity": "FAKE",
    "field": "id",
  },
  Symbol(metadata): Object {},
}
`;
